src/raytracer.cc

```diff
@@ -0,0 +1,179 @@
+#define _USE_MATH_DEFINES // Needed to run in windows/visual studio
+#include "raytracer.h"
+#include "scene.h"
+#include "scene_object.h"
+// TODO: Remove when we have all point lights in vector
+#include "point_light.h"
+#include <random>
+#include <iostream>
+
+// TODO: Place these somewhere that makes the most sense and remove some?
+const float EPSILON = 0.00001f;
+const float REFRACTION_FACTOR_OI = REFRACTION_INDEX_AIR / REFRACTION_INDEX_GLASS; // outside->in
+const float REFRACTION_FACTOR_IO = REFRACTION_INDEX_GLASS / REFRACTION_INDEX_AIR; // inside->out
+const float CRITICAL_ANGLE = asin(REFRACTION_FACTOR_OI);
+const unsigned int MAX_DEPTH = 10; // What Max depth makes sense?
+const float gamma_factor = 3.6f;
+
+std::default_random_engine generator2;
+std::uniform_real_distribution<float> distribution2(0, 1);
+
+Raytracer::Raytracer() {}
+
+ColorDbl Raytracer::CalculateDirectIllumination(Ray& ray, IntersectionPoint& p, Scene& scene) {
+  const std::vector<std::unique_ptr<Light>>& lights = scene.get_lights();
+  ColorDbl color_accumulator = COLOR_BLACK;
+  for (auto& light : lights) {
+    Direction light_direction = light->get_position() - p.get_position();
+
+    // Set dot product to zero if light is behind the surface
+    Direction unit_surface_normal = glm::normalize(p.get_normal());
+    Vertex shadow_point_origin = p.get_position() + unit_surface_normal * 0.00001f;
+
+    // Compute shadow ray
+    Ray shadow_ray = Ray(shadow_point_origin, light_direction);
+    bool in_shadow = CastShadowRay(shadow_ray, scene, light_direction);
+
+    if (!in_shadow) {
+      Direction unit_light_direction = glm::normalize(light_direction);
+      float l_dot_n = fmax(0.f, glm::dot(unit_light_direction, unit_surface_normal));
+      color_accumulator += light->get_intensity() * light->get_color() * l_dot_n;
+    }
+  }
+  return color_accumulator * p.get_material().get_color();
+}
+
+ColorDbl Raytracer::Shade(Ray& ray, IntersectionPoint& p, Scene& scene, unsigned int& depth) {
+  if (depth > MAX_DEPTH || ray.get_importance() < 0.05) {
+    return CalculateDirectIllumination(ray, p, scene);
+  }
+  if (p.get_material().get_specular() > 0.f) {
+    Direction n = glm::normalize(p.get_normal());
+    Direction d = ray.get_direction();
+
+    Vertex reflection_point_origin = p.get_position() + n * 0.00001f;
+    Direction reflection_direction = d - 2*(glm::dot(d, n))*n;
+    Ray reflection_ray = Ray(reflection_point_origin, reflection_direction);
+    reflection_ray.has_hit_diffuse = ray.has_hit_diffuse;
+    return Raytrace(reflection_ray, scene, depth + 1);
+  }
+  if (p.get_material().get_transparence() > 0.f) { // If object has refractive component
+    return p.get_material().get_color() * HandleRefraction(ray, p, scene, depth);
+  }
+
+  //STOP ON SECOND DIFFUSE HIT
+  if (ray.has_hit_diffuse) {
+    return CalculateDirectIllumination(ray, p, scene);
+  }
+  ray.has_hit_diffuse = true;
```

Comment:
Why keep this in the ray instance? Imo this should be a global or a local that we pass recursively (and have 1 instance of), instead of recreating it every time we create new Ray.